### ISSUE TITLE
[dev task] NOD | Send letter feed info to ADG team

### ISSUE THREAD

**Issue by saderagsdale** [department-of-veterans-affairs/va.gov-team#71189]
## Background Context
Part of the remediation process for the infinite loop bug, the VA would like to send letters notifying, potentially affected veterans. To make this happen, they need to set up a "letter feed". We need to gather the details for them. Email content is below:

ADG needs the following for a new feed to be setup:

- [x] Name of the feed
eFolder attributes
  - 10182 Infinite Loop Error Outreach Letter
- [x] DocType:  Identifies the type of document being uploaded via this feed. Values provided must align to the  Master Document Type Spreadsheet containing the current valid document types/taxonomy recognized by users of the eFolder.
  - Lookup ID: 184
  - Description: Notification Letter (e.g. VA 20-8993, VA 21-0290, PCGL)
- [x] Subject: Identifies the value populated in the eFolder Subject attribute for each document uploaded via this feed.
  - VA.gov 10182 Form Error Letter
- [x] Content Source: identifies the original source of the letter content. This value must align to a valid Content Source value from the current Master Document Type Spreadsheet
  - ADG
- [x] System Source: Identifies the system that uploads the file(s) to the eFolder. Normally is set to “ADG”.
  - ADG.
- [x] Source Comment: Identifies the value populated in the Source Comment attribute for each document uploaded via this feed.
  - VA.gov 10182 Form Error Letter
- [x] Letter File Name (defined by Brady based on data file received): The acceptable format of the zip file name ADG will expect to receive for this feed based on data file name. Value is case sensitive. e.g. [numeric value]_PGMemo-MMDDYYYY_*.dat.zip where * is the is the numeric value for the file.
- [x] Expected Frequency: Provides information relative to the frequency in which this ADG feed will be executed in production (daily, weekly, monthly, yearly, ad hoc, other).  If the feed will be on a recurring schedule (i.e. every 3rd Friday of the month), this information is also provided.
  - "On Demand"
- [x] Expected Quantity of Documents: Provides information on how many documents will be processed through the feed.
  - 3,152.  See https://dsva.slack.com/archives/CKUCFQAHJ/p1701621634179969?thread_ts=1701300806.930029&cid=CKUCFQAHJ.
- [x] Letter Sponsor: Identifies the Line of Business group that requested and manages the letter content and frequency of feeds.
  - Board of Veteran Appeals; John Gosnell (John.Gosnell@va.gov)

---

## Definition of Ready 

- [ ] We know to to work with to gather info (enablement team)

## Definition of Done

- [ ] We've passed details to the ADG team (see email Sade forwarded to Zach)

**Comment by anniebtran** [comment]
Leaving an update before I leave for some time off — Sent a message to the email thread, got a few OOO messages and Tracy/Mandel are both out of office for at least the next few days and in those messages, said to reach out to Dan Nguyen in their absence, so I looped him into the email thread to see if he could help us/jump on a call to chat through the items needed for ADG to create a new feed.

**Comment by saderagsdale** [comment]
Sade will check email today.

**Comment by HeatherWidmont** [comment]
We're in good shape right now, if anything changes @saderagsdale will update the ticket with a comment

**Comment by HeatherWidmont** [comment]
There are a couple of pieces Julie asked for us to get the mailing info & email address & file number for the Veterans that were affected by the infinite loop bug. @data-doge sent a huge spreadsheet to Premal previously that may hold some of that info, they can find that spreadsheet in our team sharepoint. Nice to have: Can we see Power of Attorney info?

**Comment by HeatherWidmont** [comment]
@saderagsdale to send link to the spreadsheet to @data-doge

**Comment by saderagsdale** [comment]
Eugene found it in MS Teams

**Comment by HeatherWidmont** [comment]
Still TBD on if the ADG team got what they needed. @data-doge making progress on the mailing info, email address, & file number. They are drafting a script to get the information, still some outstanding questions they are working through, though. Thomas has a way to run in one big batch & export to CSV in secure S3 bucket so once the script is ready, Eugene will reach out to Thomas.

**Comment by eileen-coforma** [comment]
Continue to work on script. If unfinished, will include in transition plan.

Note: They don't want to send out these letters until NOD v2 form is released, so launch block bugs are higher priority, but this is still prioritized

**Comment by data-doge** [comment]
Here is my working script for gathering a Veteran's file_number, email, and mailing address, given their ICN:

```
def get_data(icn)
  mpi_service = MPI::Service.new
  mpi_profile = mpi_service.find_profile_by_identifier(identifier: icn, identifier_type: 'ICN')&.profile
  raise 'No mpi profile!' if mpi_profile.nil?

  mpi_address = mpi_profile.address
  raise 'No address in mpi profile!' if mpi_address.nil?
  line_1 = mpi_address.street
  line_2 = mpi_address.street2
  street = line_1
  street += ", #{line_2}" if line_2.present?
  city = mpi_address.city
  postal_code = mpi_address.postal_code
  country = mpi_address.country

  address = "#{street}, #{city} #{postal_code}, #{country}"

  vet360_profile = VAProfile::ContactInformation::Service.get_person(mpi_profile.vet360_id.to_s)&.person
  raise 'No vet360 profile!' if vet360_profile.nil?

  emails = vet360_profile.emails
  current_emails = emails.select { |email| email.effective_end_date.nil? }
  raise 'All emails expired!' if current_emails.length == 0
  email = current_emails.first.email_address

  first_name = mpi_profile.given_names[0]
  middle_name = mpi_profile.given_names[1]
  last_name = mpi_profile.family_name
  suffix = mpi_profile.suffix
  common_name = [first_name, middle_name, last_name, suffix].compact.join(' ')

  bgs_external_key = (common_name.presence || email).first(39)
  bgs_service = BGS::Services.new(external_uid: icn, external_key: bgs_external_key)
  participant_id = mpi_profile.participant_id
  ssn = mpi_profile.ssn
  bgs_person = bgs_service.people.find_person_by_ptcpnt_id(participant_id) || bgs_service.people.find_by_ssn(ssn)
  raise 'No BGS person record found!' unless bgs_person
  file_number = bgs_person[:file_nbr]
  raise 'No file number!' unless file_number
  file_number = file_number.delete('-') if file_number =~ /\A\d{3}-\d{2}-\d{4}\z/

  [icn, file_number, email, address]
end
```

And here is a list of the ICNs for Veterans potentially impacted by the NOD infinite loop bug (i.e. the Veterans that ADG/VA will be sending letters/emails to): [Nod Infinite Loop Bug ICNs.csv](https://github.com/department-of-veterans-affairs/va.gov-team/files/13806305/Nod.Infinite.Loop.Bug.ICNs.csv)

**Comment by HeatherWidmont** [comment]
@anniebtran taking over this work while Eugene is OOO
